### ISSUE TITLE
Type narrowing fails to account for multiple conditions

### ISSUE THREAD

**Issue by ljharb** [microsoft/TypeScript#53072]
# Bug Report

### 🔎 Search Terms

type narrowing

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about _________

### ⏯ Playground Link

https://tsplay.dev/mAJ91W

### 💻 Code

```
function isNaN(x: unknown): x is number {
	return x !== x;
}
function SameValue(x: unknown, y: unknown): x is typeof y {
	if (x === y) { // 0 === -0, but they are not identical.
		if (x === 0) { return 1 / x === 1 / y; }
		return true;
	}
	return isNaN(x) && isNaN(y);
};
```

### 🙁 Actual behavior

The `y` in `1 / y` has a type error.

### 🙂 Expected behavior

Since this is inside a block where `x === y`, and also where `x === 0`, TS should be able to statically know that `x` and `y` are of type `number` (since both of them is definitively a zero, either `+0` or `-0`). I shouldn't need the `y === 0` check just to please TS.

**Comment by fatcerberus** [comment]
```ts
function isNaN(x: unknown): x is number
```

For the record, this is wrong as it implies that `x is NOT number` if the function returns `false`.  See https://github.com/microsoft/TypeScript/issues/15048.

```ts
function SameValue(x: unknown, y: unknown): x is typeof y
```

This simply expands to `x is unknown`.  You'd need to make it generic to do what this signature implies.

**Comment by ljharb** [comment]
@fatcerberus thanks, that's fair about the isNaN thing - unfortunately TS doesn't allow `is NaN`.

the `x is typeof y` isn't the important part; replace it with `boolean` and the same type error appears for this issue, but thanks for the pointer :-)

**Comment by fatcerberus** [comment]
Yeah, I figured the return signature of `SameValue` was irrelevant to the main issue, but didn't want to develop any misconceptions about how type predicates work, lest another duplicate of #15048 shows up :wink:
